File: TMA/webapp/src/crypto/bip39.ts
/**
 * BIP39 (24 words, English, SHA-256 checksum, PBKDF2 seed).
 * Logic aligned with Android `com.marsa.chain.crypto.hd.Bip39`.
 */

async function sha256(bytes: Uint8Array): Promise<Uint8Array> {
  const buf = await crypto.subtle.digest("SHA-256", bytes);
  return new Uint8Array(buf);
}

export async function loadEnglishWordList(): Promise<string[]> {
  const r = await fetch("/bip39_english.txt");
  if (!r.ok) throw new Error("Failed to load BIP39 word list");
  const list = (await r.text())
    .split(/\r?\n/)
    .map((s) => s.trim().toLowerCase())
    .filter((s) => s.length > 0);
  if (list.length !== 2048) throw new Error(`BIP39 list size ${list.length}, expected 2048`);
  return list;
}

function entropyToBits(entropy: Uint8Array, hashFirst: Uint8Array): boolean[] {
  const checksumBits = entropy.length / 4;
  const totalBits = entropy.length * 8 + checksumBits;
  const bits = new Array<boolean>(totalBits);
  for (let i = 0; i < entropy.length; i++) {
    for (let bit = 0; bit < 8; bit++) {
      bits[i * 8 + bit] = ((entropy[i]! >> (7 - bit)) & 1) === 1;
    }
  }
  for (let bit = 0; bit < checksumBits; bit++) {
    bits[entropy.length * 8 + bit] = ((hashFirst[0]! >> (7 - bit)) & 1) === 1;
  }
  return bits;
}

function bitsToMnemonic(bits: boolean[], wordList: string[]): string {
  const words: string[] = [];
  let i = 0;
  while (i < bits.length) {
    let idx = 0;
    for (let b = 0; b < 11; b++) {
      idx <<= 1;
      if (i + b < bits.length && bits[i + b]) idx |= 1;
    }
    words.push(wordList[idx] ?? "");
    i += 11;
  }
  return words.join(" ");
}

export async function generateMnemonic(wordList: string[]): Promise<string> {
  if (wordList.length !== 2048) throw new Error("Invalid word list");
  const entropy = new Uint8Array(32);
  crypto.getRandomValues(entropy);
  const hash = await sha256(entropy);
  const bits = entropyToBits(entropy, hash);
  return bitsToMnemonic(bits, wordList);
}

export async function validateMnemonicPhrase(mnemonic: string, wordList: string[]): Promise<boolean> {
  const words = mnemonic
    .trim()
    .split(/\s+/)
    .filter((w) => w.length > 0)
    .map((w) => w.toLowerCase());
  if (words.length !== 24) return false;
  const indices = new Int32Array(24);
  for (let wi = 0; wi < 24; wi++) {
    const idx = wordList.indexOf(words[wi]!);
    if (idx < 0) return false;
    indices[wi] = idx;
  }
  const concatBits = new Array<boolean>(24 * 11);
  for (let wi = 0; wi < 24; wi++) {
    const wordIndex = indices[wi]!;
    for (let bit = 0; bit < 11; bit++) {
      concatBits[wi * 11 + bit] = ((wordIndex >> (10 - bit)) & 1) === 1;
    }
  }
  const entropyBits = 256;
  const entropy = new Uint8Array(entropyBits / 8);
  for (let i = 0; i < entropyBits; i++) {
    if (concatBits[i]) {
      const bi = i >>> 3;
      entropy[bi] = (entropy[bi]! | (1 << (7 - (i % 8)))) >>> 0;
    }
  }
  const hash = await sha256(entropy);
  for (let bit = 0; bit < 8; bit++) {
    const expected = ((hash[0]! >> (7 - bit)) & 1) === 1;
    if (concatBits[entropyBits + bit] !== expected) return false;
  }
  return true;
}

export async function mnemonicToSeedBytes(mnemonic: string, passphrase = ""): Promise<Uint8Array> {
  const norm = mnemonic
    .trim()
    .split(/\s+/)
    .join(" ");
  const enc = new TextEncoder();
  const keyMaterial = await crypto.subtle.importKey("raw", enc.encode(norm.normalize("NFKD")), "PBKDF2", false, [
    "deriveBits",
  ]);
  const salt = enc.encode(`mnemonic${passphrase}`.normalize("NFKD"));
  const bits = await crypto.subtle.deriveBits(
    { name: "PBKDF2", salt, iterations: 2048, hash: "SHA-512" },
    keyMaterial,
    512,
  );
  return new Uint8Array(bits);
}

File: android-client/app/src/main/java/com/marsa/chain/crypto/hd/Bip39.kt
package com.marsa.chain.crypto.hd

import android.content.Context
import com.marsa.chain.R
import java.io.BufferedReader
import java.io.InputStreamReader
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec


object Bip39 {

    fun loadEnglishWordList(context: Context): List<String> {
        val list = context.resources.openRawResource(R.raw.bip39_english).use { ins ->
            BufferedReader(InputStreamReader(ins, Charsets.UTF_8))
                .readLines()
                .map { it.trim().lowercase() }
                .filter { it.isNotEmpty() }
        }
        require(list.size == 2048) { "word list size ${list.size}" }
        return list
    }

    
    fun loadWordListFromLines(lines: List<String>): List<String> =
        lines.map { it.trim().lowercase() }.filter { it.isNotEmpty() }
            .also { require(it.size == 2048) }

    fun generateMnemonic(wordList: List<String>): String {
        require(wordList.size == 2048)
        val entropy = ByteArray(32).also { SecureRandom().nextBytes(it) }
        return mnemonicFromEntropy(entropy, wordList)
    }

    fun mnemonicFromEntropy(entropy: ByteArray, wordList: List<String>): String {
        require(entropy.size == 32) { "256-bit entropy only for 24 words" }
        require(wordList.size == 2048)
        val hash = MessageDigest.getInstance("SHA-256").digest(entropy)
        val checksumBits = entropy.size / 4 // 8 bits
        val totalBits = entropy.size * 8 + checksumBits
        val bits = BooleanArray(totalBits)
        // entropy bits big-endian per byte? BIP39 uses first byte MSB first
        for (i in entropy.indices) {
            for (bit in 0 until 8) {
                bits[i * 8 + bit] = (entropy[i].toInt() shr (7 - bit) and 1) == 1
            }
        }
        for (bit in 0 until checksumBits) {
            bits[entropy.size * 8 + bit] = (hash[0].toInt() shr (7 - bit) and 1) == 1
        }
        val words = ArrayList<String>(24)
        var i = 0
        while (i < totalBits) {
            var idx = 0
            for (b in 0 until 11) {
                idx = idx shl 1
                if (i + b < totalBits && bits[i + b]) idx = idx or 1
            }
            words.add(wordList[idx])
            i += 11
        }
        return words.joinToString(" ")
    }

    fun validateMnemonicPhrase(mnemonic: String, wordList: List<String>): Boolean {
        val words = mnemonic.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }.map { it.lowercase() }
        if (words.size != 24) return false
        val indices = IntArray(24)
        for (w in words.withIndex()) {
            val idx = wordList.indexOf(w.value)
            if (idx < 0) return false
            indices[w.index] = idx
        }
        val concatBits = BooleanArray(24 * 11)
        for (wi in 0 until 24) {
            val wordIndex = indices[wi]
            for (bit in 0 until 11) {
                concatBits[wi * 11 + bit] = (wordIndex shr (10 - bit) and 1) == 1
            }
        }
        val entropyBits = 256
        val entropy = ByteArray(entropyBits / 8)
        for (i in 0 until entropyBits) {
            if (concatBits[i]) {
                entropy[i / 8] = (entropy[i / 8].toInt() or (1 shl (7 - (i % 8)))).toByte()
            }
        }
        val hash = MessageDigest.getInstance("SHA-256").digest(entropy)
        for (bit in 0 until 8) {
            val expected = (hash[0].toInt() shr (7 - bit) and 1) == 1
            if (concatBits[entropyBits + bit] != expected) return false
        }
        return true
    }

    fun mnemonicToSeedBytes(mnemonic: String, passphrase: String = ""): ByteArray {
        val norm = mnemonic.trim().split(Regex("\\s+")).joinToString(" ")
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512")
        val spec = PBEKeySpec(
            norm.toCharArray(),
            ("mnemonic$passphrase").toByteArray(Charsets.UTF_8),
            2048,
            64 * 8
        )
        return factory.generateSecret(spec).encoded
    }
}
